runtime/src/registrar.rs

```diff
@@ -0,0 +1,381 @@
+		initial_head_data: Vec<u8>
+	) -> Result;
+
+	/// Deregister a parachain with given `id`. If `id` is not currently registered, an error is returned.
+	fn deregister_para(id: ParaId) -> Result;
+}
+
+impl<T: Trait> Registrar<T::AccountId> for Module<T> {
+	fn new_id() -> ParaId {
+		<NextFreeId>::mutate(|n| { let r = *n; *n = ParaId::from(u32::from(*n) + 1); r })
+	}
+
+	fn register_para(
+		id: ParaId,
+		info: ParaInfo,
+		code: Vec<u8>,
+		initial_head_data: Vec<u8>
+	) -> Result {
+		if let Scheduling::Always = info.scheduling {
+			Parachains::mutate(|parachains|
+				match parachains.binary_search(&id) {
+					Ok(_) => Err("Parachain already exists"),
+					Err(idx) => {
+						parachains.insert(idx, id);
+						Ok(())
+					}
+				}
+			)?;
+		}
+		<parachains::Module<T>>::initialize_para(id, code, initial_head_data);
+		Ok(())
+	}
+
+	fn deregister_para(id: ParaId) -> Result {
+		let info = Paras::take(id).ok_or("Invalid id")?;
+		if let Scheduling::Always = info.scheduling {
+			Parachains::mutate(|parachains|
+				parachains.binary_search(&id)
+					.map(|index| parachains.remove(index))
+					.map_err(|_| "Invalid id")
+			)?;
+		}
+		<parachains::Module<T>>::cleanup_para(id);
+		Ok(())
+	}
+}
+
+type BalanceOf<T> =
+	<<T as Trait>::Currency as Currency<<T as system::Trait>::AccountId>>::Balance;
+type NegativeImbalanceOf<T> =
+	<<T as Trait>::Currency as Currency<<T as system::Trait>::AccountId>>::NegativeImbalance;
+
+pub trait Trait: parachains::Trait {
+	/// The overarching event type.
+	type Event: From<Event> + Into<<Self as system::Trait>::Event>;
+
+	/// The system's currency for parathread payment.
+	type Currency: Currency<Self::AccountId>;
+
+	/// The deposit to be paid to run a parathread.
+	type ParathreadDeposit: Get<BalanceOf<Self>>;
+
+	/// Handler for when two ParaIds are swapped.
+	type OnSwap: OnSwap;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Registrar {
+		// Ordered vector of all parachain IDs.
+		Parachains: Vec<ParaId>;
+
+		/// The number of threads to schedule per block.
+		ThreadCount: u32;
+
+		/// Current set of threads the next block; ordered by para ID. There can be no duplicates
+		/// of para ID in the list.
+		SelectedThreads: Vec<(ParaId, CollatorId)>;
+
+		/// Parathreads/chains scheduled for execution this block. If the collator ID is set, then
+		/// a particular collator has already been chosen for the next block, and no other collator
+		/// may provide the block.
+		///
+		/// Ordered as Parachains ++ Selected_Parathreads.
+		Active: Vec<(ParaId, Option<CollatorId>)>;
+
+		/// The next unused ParaId value. Start this high in order to keep low numbers for
+		/// system-level chains.
+		NextFreeId: ParaId = LOWEST_USER_ID;
+
+		/// Pending swap operations.
+		PendingSwap: map ParaId => ParaId;
+
+		// Map of all registered parathreads/chains.
+		Paras get(paras): map ParaId => Option<ParaInfo>;
+	}
+	add_extra_genesis {
+		config(parachains): Vec<(ParaId, Vec<u8>, Vec<u8>)>;
```

Comment:
Could we use type aliases (as mentioned in previous comment https://github.com/paritytech/polkadot/pull/341/files#r308993502) to change `Vec<u8>, Vec<u8>` to something like `CodeHash, InitialHeadData` (if my assumption of what they represent is correct)?